src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    //  even though that is *also* a requirement of `posix_memalign`.)
+    align: Alignment,
+}
+
+
+// FIXME: audit default implementations for overflow errors,
+// (potentially switching to overflowing_add and
+//  overflowing_mul as necessary).
+
+impl Layout {
+    pub(crate) fn from_size_align(size: usize, align: usize) -> Layout {
+        assert!(align.is_power_of_two());
+        assert!(align > 0);
+        Layout { size: size, align: align }
+    }
+
+    /// The minimum size in bytes for a memory block of this layout.
+    pub fn size(&self) -> usize { self.size }
+
+    /// The minimum byte alignment for a memory block of this layout.
+    pub fn align(&self) -> usize { self.align }
+
+    /// Constructs a `Layout` suitable for holding a value of type `T`.
+    pub fn new<T>() -> Self {
+        let (size, align) = size_align::<T>();
+        Layout::from_size_align(size, align)
+    }
+
+    /// Produces layout describing a record that could be used to
+    /// allocate backing structure for `T` (which could be a trait
+    /// or other unsized type like a slice).
+    pub fn for_value<T: ?Sized>(t: &T) -> Self {
+        let (size, align) = (mem::size_of_val(t), mem::align_of_val(t));
+        Layout::from_size_align(size, align)
+    }
+
+    /// Creates a layout describing the record that can hold a value
+    /// of the same layout as `self`, but that also is aligned to
+    /// alignment `align` (measured in bytes).
+    ///
+    /// If `self` already meets the prescribed alignment, then returns
+    /// `self`.
+    ///
+    /// Note that this method does not add any padding to the overall
+    /// size, regardless of whether the returned layout has a different
+    /// alignment. In other words, if `K` has size 16, `K.align_to(32)`
+    /// will *still* have size 16.
+    pub fn align_to(&self, align: Alignment) -> Self {
+        if align > self.align {
+            let pow2_align = align.checked_next_power_of_two().unwrap();
+            debug_assert!(pow2_align > 0); // (this follows from self.align > 0...)
+            Layout { align: pow2_align,
+                     ..*self }
+        } else {
+            self.clone()
+        }
+    }
+
+    /// Returns the amount of padding we must insert after `self`
+    /// to ensure that the following address will satisfy `align`
+    /// (measured in bytes).
+    ///
+    /// Behavior undefined if `align` is not a power-of-two.
+    ///
+    /// Note that in practice, this is only useable if `align <=
+    /// self.align` otherwise, the amount of inserted padding would
+    /// need to depend on the particular starting address for the
+    /// whole record, because `self.align` would not provide
+    /// sufficient constraint.
+    pub fn padding_needed_for(&self, align: Alignment) -> usize {
+        debug_assert!(align <= self.align());
+        let len = self.size();
+        let len_rounded_up = (len + align - 1) & !(align - 1);
```

Comment:
Arguably, this should actually be wrapping arithmetic? That will produce the right answer in all cases, even when `len + align` overflows.